lib/json_logic/base.ex

```diff
@@ -0,0 +1,652 @@
+          value = if(is_map(value), do: __MODULE__.apply(value, data), else: value)
+
+          value == 42
+        end
+      end
+
+  There are a few things to explain here:
+
+  1. We can use any module name that fits our application structure and even define multiple custom JsonLogic extensions
+     for different areas. We just need to make sure to use their respective [`apply/2`](`JsonLogic.apply/2`) function
+     when evaluating logic that needs access to our custom operations.
+  2. All custom operations must be listed in the mapping from string keys (used in JsonLogic) to function atoms defined
+     in this module. Since this mechanism allows overriding built-in operations, be carefull when chosing the name for
+     your operation (see the list of [supported operations](https://jsonlogic.com/operations.html) for identifiers
+     already in use).
+  3. For this operation that only takes one argument, we can support both versions of passing that argument: As a list
+     with one item or as a bare value. For operations with multiple arguments, adjust the match clauses appropriately.
+  4. When calling other JsonLogic operators (either through [`apply/2`](`JsonLogic.apply/2`) or directly), always use
+     `__MODULE__` or an alias pointing to **our own** JsonLogic module. When referencing the *original* `JsonLogic`
+     module, our custom operations will not be available for nested evaluations.
+
+  By following these rules, JsonLogic can be extended in a reproducible way.
+
+      iex> MyApp.JsonLogic.apply(%{"is_42" => [%{"var" => "answer"}]}, %{"answer" => 42})
+      true
+
+  """
+
+  import JsonLogic.ImplHelpers
+
+  @doc false
+  def operations(),
+    do: %{
+      "var" => :operation_var,
+      "missing" => :operation_missing,
+      "missing_some" => :operation_missing_some,
+      "if" => :operation_if,
+      "?:" => :operation_if,
+      "==" => :operation_similar,
+      "!=" => :operation_not_similar,
+      "===" => :operation_equal,
+      "!==" => :operation_not_equal,
+      "!" => :operation_not,
+      "!!" => :operation_not_not,
+      "or" => :operation_or,
+      "and" => :operation_and,
+      "<" => :operation_less_than,
+      ">" => :operation_greater_than,
+      "<=" => :operation_less_than_or_equal,
+      ">=" => :operation_greater_than_or_equal,
+      "max" => :operation_max,
+      "min" => :operation_min,
+      "+" => :operation_addition,
+      "-" => :operation_subtraction,
+      "*" => :operation_multiplication,
+      "/" => :operation_division,
+      "%" => :operation_remainder,
+      "map" => :operation_map,
+      "filter" => :operation_filter,
+      "reduce" => :operation_reduce,
+      "all" => :operation_all,
+      "none" => :operation_none,
+      "some" => :operation_some,
+      "merge" => :operation_merge,
+      "in" => :operation_in,
+      "cat" => :operation_cat,
+      "substr" => :operation_substr,
+      "log" => :operation_log
+    }
+
+  defmacro __using__(opts) do
```

Comment:
Giant macros like this aren't a good idea for long term maintenance.

A better solution would be to utilize something like a protocol system where by an extension could be layered in on top of the base. 

I experimented with this by using a middleware approach, but it got a little complicated and I accidentally created a case where it would go into an infinite recursion loop.

Reply:
How I envisioned this was to do something like

```elixir
defmodule MyApp.JsonLogic do
  use JsonLogic, 
    extensions: [
      MyApp.JsonLogic.CustomOperation,
      MyApp.JsonLogic.OtherCustomThing,
    ]
end
```

The extensions would then implement a standard interface and if it was an unsupported operation they _could_ return `:ignore` or `:unsupported` and the json logic module that was used, would fall through to the next extension in line and once it has exhausted all of the steps, it would fail.

The one problem I faced initially with this approach was applying the logic on a recursive basis. You'd have to figure out how to call back to the top level again to start the search in the stack all over.

Reply:
I think this approach would work if the list of extensions happened in the config file:

```elixir
config :json_logic, extensions: [
  MyApp.JsonLogic.CustomOperation
]
```

This way, there is a global register to loop up registered extensions for JsonLogic.

The downsides are:
- This pattern is discouraged (see https://hexdocs.pm/elixir/main/library-guidelines.html#avoid-application-configuration. Ecto does a similar thing, but scoped to a specific `MyApp.Repo`. But if you need the module name to lookup the config, that's no different to looking it up on the module itself)
- We would limit the number of separately configured JsonLogic "instances"